### PR TITLE
Add a new way to start speech server

### PR DESCRIPTION
**Pull request by xutaima**
Add a new way to start speech  server, which might be easier for participants who didn't run the baseline experiments or is working on other framework.

The new one needs two files
- TGT_FILE: the file with reference sentences
- WAV_LIST_FILE: the file with a list of paths to WAVs, line aligned to TGT_FILE 

In this case we can start the server
```shell
python $fairseq/example/simultaneous_translation/eval/server.py \
    --tokenizer 13a \
    --src-file SRC_FILE \
    --tgt-file WAV_LIST_FILE \
    --tgt-file-type text \
    --scorer-type speech \
    --output $result_dir \
    --port 12321
```

### CODE REVIEW

File: examples/simultaneous_translation/docs/evaluation.md [hunk 1]
```diff
@@ -5,22 +5,50 @@ The server provides information needed by the client and evaluates latency and q
 We use the Fairseq toolkit as an example but the evaluation process can be applied in an arbitary framework.
 
 ## Server
-The server code is provided and can be set up locally for development purposes. For example, to evaluate a text simultaneous test set,
+The server code is provided and can be set up locally for development purposes. 
+Server sends source words to client, and record the delay when receving the prediction.
```
Comment (jmp84) [inline examples/simultaneous_translation/docs/evaluation.md:9]:
```suggestion
The server sends source words or speech segments to the client, and records the delay when receiving predictions.
```

File: examples/simultaneous_translation/docs/evaluation.md [hunk 2]
```diff
@@ -5,22 +5,50 @@ The server provides information needed by the client and evaluates latency and q
...
+Here is an instruction on how to setup server for development.
```
Comment (jmp84) [inline examples/simultaneous_translation/docs/evaluation.md:10]:
```suggestion
Here are instructions on how to setup the server for development.
```

File: examples/simultaneous_translation/docs/evaluation.md [hunk 3]
```diff
@@ -5,22 +5,50 @@ The server provides information needed by the client and evaluates latency and q
...
+
+To run start a text server listening at port 12321, with `SRC_FILE` and `TGT_FILE` as raw source and target text.
```
Comment (jmp84) [inline examples/simultaneous_translation/docs/evaluation.md:12]:
```suggestion
To run start a text server listening at port 12321, with `$SRC_FILE` and `$TGT_FILE` as raw source and target text and `$result_dir` a directory to store the results:
```

File: examples/simultaneous_translation/docs/evaluation.md [hunk 4]
```diff
@@ -5,22 +5,50 @@ The server provides information needed by the client and evaluates latency and q
...
+
+```shell
+python $fairseq/example/simultaneous_translation/eval/server.py \
+    --tokenizer 13a \
+    --src-file SRC_FILE \
```
Comment (jmp84) [inline examples/simultaneous_translation/docs/evaluation.md:17]:
```suggestion
    --src-file $SRC_FILE \
```

File: examples/simultaneous_translation/docs/evaluation.md [hunk 5]
```diff
@@ -5,22 +5,50 @@ The server provides information needed by the client and evaluates latency and q
...
+    --tgt-file TGT_FILE \
```
Comment (jmp84) [inline examples/simultaneous_translation/docs/evaluation.md:18]:
```suggestion
    --tgt-file $TGT_FILE \
```

File: examples/simultaneous_translation/docs/evaluation.md [hunk 6]
```diff
@@ -5,22 +5,50 @@ The server provides information needed by the client and evaluates latency and q
...
+    --scorer-type text \
+    --output $result_dir \
+    --port 12321
+```
+
+As to speech models, if you have gone through the Data Preparation in [baseline experiment](baseline.md), you can use the json file in $DATA_ROOT/data-bin/mustc_en_de, for example, dev.json. So we can start the server
```
Comment (jmp84) [inline examples/simultaneous_translation/docs/evaluation.md:24]:
```suggestion
For speech models, if you have gone through the Data Preparation in [baseline experiment](baseline.md), you can use the json file in $DATA_ROOT/data-bin/mustc_en_de, for example, dev.json. So we can start the server:
```

File: examples/simultaneous_translation/docs/evaluation.md [hunk 7]
```diff
@@ -5,22 +5,50 @@ The server provides information needed by the client and evaluates latency and q
...
 ```shell
-python $user_dir/eval/server.py \
+python $fairseq/example/simultaneous_translation/eval/server.py \
     --tokenizer 13a \
-    --src-file $src \
-    --tgt-file $tgt \
-    --scorer-type {text, speech} \
-    --output $result_dir/eval \
+    --tgt-file $DATA_ROOT/data-bin/mustc_en_de/dev.json \
+    --tgt-file-type json \
```
Comment (jmp84) [inline examples/simultaneous_translation/docs/evaluation.md:30]: This is a new option? Did you make it backward compatible with the default being the json?

File: examples/simultaneous_translation/docs/evaluation.md [hunk 8]
```diff
@@ -5,22 +5,50 @@ The server provides information needed by the client and evaluates latency and q
...
+    --scorer-type speech \
+    --output $result_dir \
     --port 12321
 ```
-The `--score-type` can be either `text` or `speech` to evaluation different tasks.
 
-For text models, the `$src` and `$tgt` are the raw source and target text.
+If you don't want to go though Data Preparation, you need to prepare two files:
+- TGT_FILE: the file with reference sentences
```
Comment (jmp84) [inline examples/simultaneous_translation/docs/evaluation.md:37]:
```suggestion
- `$TGT_FILE`: the file with reference sentences
```

File: examples/simultaneous_translation/docs/evaluation.md [hunk 9]
```diff
@@ -5,22 +5,50 @@ The server provides information needed by the client and evaluates latency and q
...
+- WAV_LIST_FILE: the file with a list of paths to WAVs, line aligned to TGT_FILE
```
Comment (jmp84) [inline examples/simultaneous_translation/docs/evaluation.md:38]:
```suggestion
- `$WAV_LIST_FILE`: the file with a list of paths to WAVs, line aligned to TGT_FILE 
```

File: examples/simultaneous_translation/docs/evaluation.md [hunk 10]
```diff
@@ -5,22 +5,50 @@ The server provides information needed by the client and evaluates latency and q
...
-For text models, the `$src` and `$tgt` are the raw source and target text.
+If you don't want to go though Data Preparation, you need to prepare two files:
+- TGT_FILE: the file with reference sentences
+- WAV_LIST_FILE: the file with a list of paths to WAVs, line aligned to TGT_FILE 
+
+In this case we can start the server
+```shell
+python $fairseq/example/simultaneous_translation/eval/server.py \
+    --tokenizer 13a \
+    --src-file SRC_FILE \
```
Comment (jmp84) [inline examples/simultaneous_translation/docs/evaluation.md:44]: Can you double check this? Where is `$TGT_FILE` ?

File: examples/simultaneous_translation/eval/server.py [hunk 11]
```diff
@@ -68,6 +69,9 @@ def add_args():
                         help='Type of data to evaluate')
     parser.add_argument('--tokenizer', default="13a", choices=["none", "13a"],
                         help='Type of data to evaluate')
+    parser.add_argument('--tgt-file-type', type=str, default="json",
```
Comment (jmp84) [inline examples/simultaneous_translation/eval/server.py:72]: Do you need to add `required=False`  just in case?

File: examples/simultaneous_translation/eval/server.py [hunk 12]
```diff
@@ -10,6 +10,7 @@
 from collections import defaultdict
 from tornado import web, ioloop
 from scorers import build_scorer
+from utils.registry import REGISTRIES
```
Comment (jmp84) [inline examples/simultaneous_translation/eval/server.py:13]: Why is this needed here?

File: examples/simultaneous_translation/docs/evaluation.md (hunk 10 above)
Comment (jmp84) [inline examples/simultaneous_translation/docs/evaluation.md:44]: I think this should be `--src-file $WAV_LIST_FILE`

File: examples/simultaneous_translation/docs/evaluation.md [hunk 13]
```diff
@@ -5,22 +5,50 @@ The server provides information needed by the client and evaluates latency and q
...
+    --tgt-file WAV_LIST_FILE \
```
Comment (jmp84) [inline examples/simultaneous_translation/docs/evaluation.md:45]: I think this should be `--tgt-file $TGT_FILE`

File: examples/simultaneous_translation/eval/scorers/speech_scorers.py [hunk 14]
```diff
@@ -9,15 +9,31 @@
...
+                    f"tgt_file type: {args.tgt_file_type}. "
+                    f"src_file {args.src_file} will be ignored.\n"
+            )
+            self.data = {
+                "src" : self._load_wav_info_from_json(args.tgt_file),
+                "tgt" : self._load_text_from_json(args.tgt_file)
+            }
+        elif args.tgt_file_type == "text": 
+            assert args.src_file is not None, (
+                "--src-file is needed if the --tgt-file is text\n"
```
Comment (jmp84) [inline examples/simultaneous_translation/eval/scorers/speech_scorers.py:24]:
```suggestion
                "--src-file is needed if --tgt-file-type is text\n"
```

File: examples/simultaneous_translation/eval/scorers/speech_scorers.py [hunk 15]
```diff
@@ -46,6 +62,12 @@ def send_src(self, sent_id, value):
             self.data["src"][sent_id]["segments"] = self._load_audio_from_path(
                 self.data["src"][sent_id]["path"]
             ) 
+            if "length" not in self.data["src"][sent_id]:
```
Comment (jmp84) [inline examples/simultaneous_translation/eval/scorers/speech_scorers.py:65]: Is there a `sent_id` available in text mode?

File: examples/simultaneous_translation/docs/evaluation.md (hunk 7 above)
Comment (xutaima) [reply examples/simultaneous_translation/docs/evaluation.md:30]: Yes the default now is json to be compatible with old code base

**Comment by xutaima** [comment]
@jmp84 Yes this pull request has be tested
If you would like to test locally, on must-c dataset your could firstly run
```bash
python $fairseq/examples/simultaneous_translation/data/yaml_to_wav_path.py $must_c_path/dev/txt/dev.yaml > dev.wav_list
```
Then start the server

```bash
python $fairseq/example/simultaneous_translation/eval/server.py \
    --tokenizer 13a \
    --src-file  dev.wav_list\
    --tgt-file $must_c_path/dev/txt/dev.de \
    --tgt-file-type text \
    --scorer-type speech \
    --port 12321
```
